### PR TITLE
Feature/load config from server

### PR DESCRIPTION
**Pull request by bumi**
This allows the user to configure the widget on the server and load the config from there. (additionally to the local props)

ref #84

### CODE REVIEW

**Comment by bumi** [comment]
(ups, there is an old commit in this PR)

File: src/containers/widget/index.js [hunk 1]
```diff
@@ -45,12 +47,23 @@ class Widget extends Component {
   }
 
   componentDidMount() {
-    this.setDefaults();
+    return fetch(`${this.props.apiBaseUrl}/v1/configuration`,
+        { headers: {"Api-Token": this.props.apiToken} })
+      .then(response => {
+        if(response.ok) {
+          response.json().then(config => {
+            this.widgetConfig = config;
```
Comment (reznord) [inline src/containers/widget/index.js:55]: Why are we storing it in a class variable here?

File: src/containers/widget/index.js [hunk 2]
```diff
@@ -45,12 +47,23 @@ class Widget extends Component {
...
+            this.setDefaults(config);
```
Comment (reznord) [inline src/containers/widget/index.js:56]: you are not consuming the config on `setDefaults`, maybe you should?

File: src/containers/widget/index.js [hunk 3]
```diff
@@ -45,12 +47,23 @@ class Widget extends Component {
...
+        } else {
+          this.setDefaults();
+        }
+      });
   }
 
   setDefaults = () => {
     this.setState({
-      ...getInitialState(this.props),
+      ...getInitialState(this.props, this.widgetConfig),
```
Comment (reznord) [inline src/containers/widget/index.js:66]: you can take the config as an argument and then pass it to getInitialState function right?

**Review by reznord — CHANGES_REQUESTED** [review]

File: src/containers/widget/index.js (hunk 3 above)
Comment (bumi) [reply src/containers/widget/index.js:66]: yes, but not from the other setDefaults() calls.

Comment (reznord) [reply src/containers/widget/index.js:66]: yeah, in that case on getInitialState, you can do something like:

```
function getInitialState(props = {}) {
  return Object.assign(Widget.defaultProps, props);
}
```

When you don't pass any argument then the default value will be an empty object.

File: src/containers/widget/index.js (hunk 2 above)
Comment (bumi) [reply src/containers/widget/index.js:56]: thanks. yes, that's old.

File: src/containers/widget/index.js (hunk 1 above)
Comment (bumi) [reply src/containers/widget/index.js:55]: to be able to set the initial state again (e.g. when the widget is closed) 
any idea how we else should do that?

File: src/containers/widget/index.js (hunk 3 above)
Comment (bumi) [reply src/containers/widget/index.js:66]: that is what is done here: https://github.com/bumi/area402-widget/pull/85/files#diff-128edf9b24e12b671c98d03163f11c4380faf0f88888c2c76a909cc786e167f2R22

still the question is what to do in the `componentWillUnmount` case? there we also call `setDefaults`

**Comment by bumi** [comment]
merging this as we are experimenting with the embedding

File: src/containers/widget/index.js (hunk 1 above)
Comment (reznord) [reply src/containers/widget/index.js:55]: Whatever the code you are putting it in componentDidMount, that will be executed every time when that component **appears** on the browser.

So, there's no point in storing the response in this.widgetConfig unless you are adding a null check on componentDidMount before making an API call.

File: src/containers/widget/index.js (hunk 3 above)
Comment (reznord) [reply src/containers/widget/index.js:66]: in `componentWillUnmount`, you can call setDefaults with `Widget.defaultProps`

File: src/containers/widget/index.js (hunk 1 above)
Comment (bumi) [reply src/containers/widget/index.js:55]: can I execute something before it appears? 

I want the call to be made once and then it should be used similar to the props (set in the HTML).

File: src/containers/widget/index.js (hunk 3 above)
Comment (bumi) [reply src/containers/widget/index.js:66]: that would remove all  settings when the modal is opened a second time.

**Pull request merged**